### PR TITLE
Update third_party/brotli subtree 

### PR DESCRIPTION
**Pull request by sherryzy**
b/277591845

### CODE REVIEW

File: third_party/brotli/c/enc/ringbuffer.h
Comment (kaidokert) [inline third_party/brotli/c/enc/ringbuffer.h:13]: This looks like it's not necessary anymore, we can just use memcpy

File: third_party/brotli/c/common/platform.h
Comment (kaidokert) [inline third_party/brotli/c/common/platform.h:27]: This should not be required anymore, we removed Starboardized `memcpy`

File: third_party/brotli/c/common/platform.c
Comment (kaidokert) [inline third_party/brotli/c/common/platform.c]: @y4vor do we still need this ? E.g. does something break if we include stdlib ?

File: third_party/brotli/c/dec/bit_reader.h
Comment (kaidokert) [inline third_party/brotli/c/dec/bit_reader.h:13]: Likewise, `memcpy` customization shouldn't be needed anymore

File: third_party/brotli/c/dec/huffman.c
Comment (kaidokert) [inline third_party/brotli/c/dec/huffman.c]: Lets leave it in the original place, if needed bracket it with #if !defined(STARBOARD)

But i don't think excluding is is needed even anymore.

File: third_party/brotli/c/dec/huffman.c
Comment (kaidokert) [inline third_party/brotli/c/dec/huffman.c]: @y4vor do we need string_poem anywhere still ? It only aliases `strdup` to Starboard currently, perhaps it's better to localize this where it's needed.

File: third_party/brotli/c/enc/hash_longest_match_quickly_inc.h
Comment (kaidokert) [inline third_party/brotli/c/enc/hash_longest_match_quickly_inc.h:251]: Shouldn't be needed anymore ?

File: third_party/brotli/c/enc/memory.h
Comment (kaidokert) [inline third_party/brotli/c/enc/memory.h:89]: No need to starboardize memcpy anymore

File: third_party/brotli/c/tools/brotli.c
Comment (kaidokert) [inline third_party/brotli/c/tools/brotli.c]: Likewise as above, we should be able to leave stdlib/string untouched

File: third_party/brotli/c/common/platform.h
Comment (dahlstrom-g) [inline third_party/brotli/c/common/platform.h:320]: Should the actual parameters here change from (p, &v, sizeof v) to (&t, p, sizeof t)?

Comment (sherryzy) [reply third_party/brotli/c/common/platform.h:320]: Oops, should be (p, &v, sizeof v). Great catch!

File: third_party/brotli/c/enc/hash_longest_match_quickly_inc.h
Comment (sherryzy) [reply third_party/brotli/c/enc/hash_longest_match_quickly_inc.h:251]: This one is added here on purpose: https://lbshell-internal-review.git.corp.google.com/c/third_party/brotli/+/99350. I guess it's still needed. cc @yuying-y to confirm.

File: third_party/brotli/c/enc/memory.h
Comment (dahlstrom-g) [inline third_party/brotli/c/enc/memory.h:88]: The backslash at EOL needs 7 more spaces indentation to match upstream commit e61745a6.

File: third_party/brotli/c/common/platform.c
Comment (dahlstrom-g) [reply third_party/brotli/c/common/platform.c]: Just to note this file (platform.c) does not exist at all on the main branch:
github.com/youtube/cobalt/tree/main/third_party/brotli/c/common

File: .gitmodules
Comment (sherryzy) [inline .gitmodules:1]: @kaidokert I'm not sure I'm doing the right thing here. As the path here should be the path to root dir, if I don't add this file, all tests fail for no url for this submodule. But all third_party code have .gitmodules file in the sub dir under third_party. Do you know how do usually we deal with this case?

File: third_party/brotli/c/dec/huffman.c
Comment (y4vor) [reply third_party/brotli/c/dec/huffman.c]: If we don't use `strdup` we don't need it in this file.

File: third_party/brotli/c/common/platform.c
Comment (y4vor) [reply third_party/brotli/c/common/platform.c]: We can try to bring it back. If it builds and there is no API leak it is fine.

File: third_party/brotli/c/enc/memory.c
Comment (dahlstrom-g) [inline third_party/brotli/c/enc/memory.c:13]: Restore to match the upstream commit?

File: third_party/brotli/c/common/platform.c
Comment (sherryzy) [reply third_party/brotli/c/common/platform.c]: I run the internal build, and there is no API leakage. (See my CL in gerrit.) But in this file, I see malloc is being used. @y4vor Could you confirm I'm doing it right?

Comment (sherryzy) [reply third_party/brotli/c/common/platform.c]: Discussed this with @y4vor offline. So maybe the code is not being used in Cobalt, so no API leakage introduced. We can just use this standard library to keep the minimum change in third-party library.

**Review by dahlstrom-g — APPROVED** [review]
I'm all out of nits!
(I'm selecting “Approve” but I'm new here and you may want to await approval from someone better acquainted with Cobalt.)

**Review by dahlstrom-g — APPROVED** [review]
I'm all out of nits!
(I'm selecting “Approve” but I'm new here and you may want to await approval from someone better acquainted with Cobalt.)

File: .gitmodules
Comment (kaidokert) [reply .gitmodules]: Sorry to hold this up - i've gone ahead and cloned your setup separately in sandbox to investigate _why_ this is happening, it should not. ( https://github.com/youtube/cobalt_sandbox/pull/437 )
 
It looks like there's another patch missing from the checkout actions code. Meanwhile - feel free to restore it in a state so that checks run, i'll separately try to resolve this issue in the sandbox

Comment (kaidokert) [reply .gitmodules]: Ok i found whats wrong, PR up at #406. You can cherry-pick that fix in to the PR if rebase is too much overhead.

File: third_party/brotli/c/enc/hash_longest_match_quickly_inc.h
Comment (kaidokert) [reply third_party/brotli/c/enc/hash_longest_match_quickly_inc.h:251]: If you remove it and builds don't issue a warning, it's okay to take it out. It may have been done against an older MSVC compiler. It's fairly minor though and can be left in to get this landed.

Comment (sherryzy) [reply third_party/brotli/c/enc/hash_longest_match_quickly_inc.h:251]: Already cherry-pick the submodule commit and see this. Maybe get this PR land first.

**Review by kaidokert — APPROVED** [review]
Looks ready to merge now - please give me a ping to confirm it should go in and i'll go ahead and do the merge.

**Review by dahlstrom-g — COMMENTED** [review]
We don't want to submit commit 028abab0 here, do we?

**Comment by dahlstrom-g** [comment]
1. We don't want to submit commit 028abab0 here, do we?

2. I took the `third_party/brotli` tree from commit 5d0e0a0a
and dropped it into a merge commit atop a simpler squash setup:
http://github.com/dahlstrom-g/cobalt/commits/brotli_update
The merge commit preserves local history in `BUILD.gn`, `METADATA`,
and elsewhere unlike the way 9b89a518 deletes all files
creating a discontinuity in history.
Feel free to grab my commits and use them here if you like.

**Comment by kaidokert** [comment]
> 1. We don't want to submit commit [028abab](https://github.com/youtube/cobalt/commit/028abab0e2a9051611d80f1dbe229557ad8f1fca) here, do we?

I'm actually fine submitting this - it's a small crutch to just verify the checks are green. It'll get elided at the time of merge because that commit is already on main. We did the same to get Jelle's PR landed here https://github.com/youtube/cobalt/pull/381

As we are in a bit of time crunch to get the updates in before a major branch cut i'm fine leaving this whole thing as is for merge.

**Pull request merged**
